[13.0][MIG] stock_operating_unit

BT-nstuder opened the pull request:
## Proposed changes

I've migrated the module `stock_operating_unit` and it's functionalities from Odoo `12.0` to `13.0`. I've checked if the all unit tests are still running.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Migration Update

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst) doc
- [x] I have signed the [CLA](https://odoo-community.org/page/cla)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated necessary documentation

## Further comments

~~This module depends on the `account_operating_unit` module. That means that there should be a migration of it to `13.0`. Because there is already a PR for a migration I've added the changes of the PR to my local environment to test my migration.~~

~~This PR depends on the current migration PR #229.~~

> I've changed the dependency of the module. Instead of `account_operating_unit` the module depends now on the module `operating_unit`

JordiBForgeFlow committed:
stock_operating_unit

JordiBForgeFlow committed:
flake8 issues

JordiBForgeFlow committed:
flake 8

JordiBForgeFlow committed:
flake 8

AaronHForgeFlow committed:
update contributors

AaronHForgeFlow committed:
code style

AaronHForgeFlow committed:
constraint that checks the OU in the picking matches the one in the move

JordiBForgeFlow committed:
the operating unit in pickings should derive from the picking type

AaronHForgeFlow committed:
travis issues

AaronHForgeFlow committed:
travis issues and test stock picking

AaronHForgeFlow committed:
test stock security according picking 1 in test_stock_picking

AaronHForgeFlow committed:
Update operating units of previous locations

AaronHForgeFlow committed:
lint check operating_unit_id readonly in the picking

Commit:
Correct error message and xml files

* Replace <openerp> and <data> elements by <odoo>
* Remove spaces in strings introduced because of \
* Remove exclamation points in error messages
* Remove most of the \ where it can be replaced by parenthesis

AaronHForgeFlow committed:
[MIG] stock_operating_unit to v10.0

esthermm committed:
[FIX] stock_operating_unit: correct constrain function, avoid error on pickings

AaronHForgeFlow committed:
[FIX]onchange does not work with readonly fields

AaronHForgeFlow committed:
[FIX]stock.move._check_stock_move_operating_unit

AaronHForgeFlow committed:
[FIX]internal locations should consider all locations but customer or suppliers

AaronHForgeFlow committed:
[FIX]field that triggers constraint is not a valid field name

bjeficent committed:
[MIG] stock_operating_unit: Migration to v12

oca-travis committed:
[UPD] Update stock_operating_unit.pot

OCA-git-bot committed:
[UPD] README.rst

MiquelRForgeFlow committed:
[FIX] Alignments

BT-nstuder committed:
[IMP] stock_operating_unit: black, isort, prettier

AaronHForgeFlow commented on stock_operating_unit/view/stock.xml:
```diff
@@ -156,24 +156,6 @@
         </field>
     </record>
 
-    <record id="view_move_picking_form" model="ir.ui.view">
```
Thanks for migrating this.

Is the operating unit in stock move not visible anymore?

BT-nstuder replied:
The operating units are still visible in the views of `stock.move`. I removed this part of code because `stock.view_move_picking_form` doesn't exist anymore.

AaronHForgeFlow replied:
All right. Thanks for answering. I will wait for #229 to review this.

alan196 commented on stock_operating_unit/model/stock_warehouse.py:
```diff
@@ -0,0 +1,59 @@
+        if self.company_id:
+            company = self.company_id
+        else:
+            company = self.env["res.company"]._company_default_get("stock.inventory")
```
The method '_company_default_get' on res.company is deprecated and shouldn't be used anymore

it's better to use self.env.company

BT-nstuder replied:
Oh, thanks for the info. I didn't know that.
Unfortunately I can't update my PR until the 03.03.2020. I will correct it then.

alan196 replied:
I will make some really heavy tests of this module with a customer with 29 operating units, I already detected some errors in this module, is there a way to take your PR and make a PR to your branch? or I make another PR with your tests and the fixes to bugs I have found?

BT-nstuder replied:
That would be great! You could add my fork as remote, create a branch out of my migration branch and afterwards make a PR to my fork. The second option would be to take my code, update it on your branch and create a new PR.

alan196 commented:
I had the following error.

### Steps to reproduce

- Create a stock.inventory for a product
- Create 3 stock.inventory.line each line with different locations of warehouses with different ou each.
- Validate Inventory

```python
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/home/odoo/instance/odoo/odoo/http.py", line 619, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/odoo/instance/odoo/odoo/http.py", line 309, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/home/odoo/instance/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/home/odoo/instance/odoo/odoo/http.py", line 664, in dispatch
    result = self._call_function(**self.params)
  File "/home/odoo/instance/odoo/odoo/http.py", line 345, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/home/odoo/instance/odoo/odoo/service/model.py", line 93, in wrapper
    return f(dbname, *args, **kwargs)
  File "/home/odoo/instance/odoo/odoo/http.py", line 338, in checked_call
    result = self.endpoint(*a, **kw)
  File "/home/odoo/instance/odoo/odoo/http.py", line 910, in __call__
    return self.method(*args, **kw)
  File "/home/odoo/instance/odoo/odoo/http.py", line 510, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/instance/odoo/addons/web/controllers/main.py", line 1320, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/home/odoo/instance/odoo/addons/web/controllers/main.py", line 1312, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/home/odoo/instance/odoo/odoo/api.py", line 395, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/home/odoo/instance/odoo/odoo/api.py", line 382, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/odoo/instance/odoo/addons/stock/models/stock_inventory.py", line 107, in action_validate
    self._action_done()
  File "/home/odoo/instance/odoo/addons/stock/models/stock_inventory.py", line 118, in _action_done
    self.post_inventory()
  File "/home/odoo/instance/odoo/addons/stock_account/models/stock_inventory.py", line 43, in post_inventory
    super(StockInventory, other_inventories).post_inventory()
  File "/home/odoo/instance/odoo/addons/stock/models/stock_inventory.py", line 125, in post_inventory
    self.mapped('move_ids').filtered(lambda move: move.state != 'done')._action_done()
  File "/home/odoo/instance/extra_addons/operating-unit/stock_account_operating_unit/model/stock_move.py", line 119, in _action_done
    "stock_move_id": self.id,
  File "/home/odoo/instance/odoo/odoo/fields.py", line 3524, in __get__
    raise ValueError("Expected singleton: %s" % record)
ValueError: Expected singleton: stock.move(5, 6, 7)
```

BT-nstuder commented:
> I had the following error.
> ### Steps to reproduce
> 
>     * Create a stock.inventory for a product
> 
>     * Create 3 stock.inventory.line each line with different locations of warehouses with different ou each.
> 
>     * Validate Inventory
> ...

Hello @alan196, I finally tried to replicate this error but I couldn't. I haven't installed the module `stock_account_operating_unit`, because this migration doesn't depend on it. Could you check again or am I missing something?

I did the following in an empty database:
1. Install module `stock_operating_unit`
2. Create 3 independent Operating Units (OU): `OU1`, `OU2`, `OU3`
3. Enable `Storage Locations` and `Multi-Warehouses` in the settings
4. Create a warehouse for each created OU and assign the correct OU: `WH1`, `WH2`, `WH3`
5. Create a location for each warehouse with the view location of the warehouse as parent: `LOC1`, `LOC2`, `LOC3`
6. Create an inventory adjustment (`stock.inventory`) for a product
7. Add 3 `stock.inventory.line`'s and each with a different location out of the 3 created before
8. Validate the inventory

No errors occurred.

ps-tubtim commented:
@BT-nstuder Please update your pr for fix conflict files.

ps-tubtim commented on stock_operating_unit/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+    "Odoo Community Association (OCA)",
+    "license": "LGPL-3",
+    "website": "https://github.com/OCA/operating-unit",
+    "depends": ["stock", "account_operating_unit"],
```
```suggestion
    "depends": ["stock", "operating_unit"],
```
IMO, I think `operating_unit` is enough. What do you think about it?

BT-nstuder replied:
I think so too. Updated it according to your comment. Thanks!

ps-tubtim commented on stock_operating_unit/readme/CONTRIBUTORS.rst:
```diff
@@ -0,0 +1,3 @@
+* Jordi Ballester Alomar <jordi.ballester@eficent.com>
+* Aaron Henriquez <aheficent@eficent.com>
+* Sudhir Arya <sudhir.arya@serpentcs.com>
```
Please add you as contributor.

ps-tubtim reviewed (commented):
Some comments.

kittiu commented:
@BT-nstuder you need to rebase to fix the conflicting files.

newtratip commented on stock_operating_unit/view/stock.xml:
```diff
@@ -0,0 +1,196 @@
+                <field name="operating_unit_id" options="{'no_create': True}"
+                       domain = "[('company_id','=', company_id),
+                       ('user_ids', 'in', uid)]"
+                       groups="operating_unit.group_multi_operating_unit"/>
```
IMO, Operating Unit should show in `stock.picking` no `stock.picking.line`

![Selection_122](https://user-images.githubusercontent.com/24691983/78154492-854b1d00-7466-11ea-988e-add79fc611db.png)

This is should be

![Selection_123](https://user-images.githubusercontent.com/24691983/78155079-3651b780-7467-11ea-904b-c0ab44b8ae05.png)

This is my code.

![Selection_124](https://user-images.githubusercontent.com/24691983/78155174-52edef80-7467-11ea-8732-c0906a4c81bf.png)

BT-nstuder replied:
You're totally right. At least in the version 12.0 it as you've suggested. Thanks!

newtratip reviewed (commented):
Reviewed.

BT-nstuder committed:
[MIG] stock_operating_unit: Migration to 13.0

BT-nstuder commented:
@ps-tubtim @kittiu @newtratip 
I've rebased my branch to solve the conflicts and added the new `prettier` tool for the XML formatting. All of your suggestions got implemented. Many thanks.

AaronHForgeFlow reviewed (commented):
I tested on runbot and it works correctly. Thank you!

Could you include those improcements from v12?
https://github.com/OCA/operating-unit/pull/267
https://github.com/OCA/operating-unit/pull/265

BT-nstuder committed:
[IMP] stock_operating_unit: Update record rule

BT-nstuder committed:
[FIX] stock_operating_unit: Constraint on stock move

BT-nstuder commented:
@AaronHForgeFlow done. Should I squash the new commits into d050dfc?

AaronHForgeFlow commented:
@BT-nstuder Thanks. No, keep them separately, better for traceability.

AaronHForgeFlow reviewed (approved):
LGTM Functional test. 

Just one thing. The setup folder is created automatically by a bot. I don't think that change regarding setup folder should be here.

BT-nstuder commented:
@AaronHForgeFlow The setup folder was created automatically when running `pre-commit run -a`. But I can still remove it if you want.

AaronHForgeFlow commented:
@BT-nstuder Yes you are correct, never mind. I was confused thinking about version 12.0. Sorry.

ps-tubtim reviewed (approved):
LGTM

newtratip commented:
@BT-nstuder codecov/project and codecov/patch still be red. I think test script is not cover.

ps-tubtim commented:
**Process:**
1. Install stock_operating_unit 
2. Create 2 OU -> OU1, OU2
3. Create 2 users for each OU -> ou1, ou2
4. Set OU in warehouses and locations
5. ou1 create a receipt.
6. ou2 **can't** see _Stock Moves_ of other OU but **can** see _Inventory Report_, _Forecasted Report_ and _Product Moves_ of other OU.

**Stock Moves**
![2020-04-10_14-52_1](https://user-images.githubusercontent.com/51266019/78873729-82a78380-7a75-11ea-98da-5ba3ec4a0bac.png)

**Inventory Report**
![2020-04-10_14-55](https://user-images.githubusercontent.com/51266019/78873766-90f59f80-7a75-11ea-87f5-52dc382a0b2f.png)

**Forecasted Report**
![2020-04-10_15-14](https://user-images.githubusercontent.com/51266019/78873784-98b54400-7a75-11ea-9435-a2f4b1e353f6.png)

**Product Moves**
![2020-04-10_14-52](https://user-images.githubusercontent.com/51266019/78873802-a074e880-7a75-11ea-89ce-f0202e0b760f.png)

@All Should users see reports of other OU?
cc: @kittiu @newtratip

AaronHForgeFlow commented:
@ps-tubtim I think not, they should not see that data in the report. We should include access rules for stock.move.line. Also inherit the reports to filter OU information accordingly.

alan196 reviewed (approved):
Functional Review 👍

AaronHForgeFlow commented on stock_operating_unit/security/stock_security.xml:
```diff
@@ -0,0 +1,91 @@
+        <field eval="1" name="perm_read" />
+        <field eval="0" name="perm_create" />
+    </record>
+    <record id="ir_rule_stock_move_allowed_operating_units" model="ir.rule">
```
could you include a similar rule for stock.move.line? That way we will cover the case for Product Moves. With that I think  this is ok to merge

AaronHForgeFlow replied:
> could you include a similar rule for stock.move.line? That way we will cover the case for Product Moves. With that I think this is ok to merge

I just created a PR to your branch with the changes. https://github.com/BT-nstuder/operating-unit/pull/1 Could you check it? If you agree with the merge, please use the option "Rebase and Merge". That way github will not generate a merge commit. Thank you!

BT-nstuder replied:
Thank you very much. I've merged your PR. Sorry for responding so late.

AaronHForgeFlow committed:
[IMP]add operating unit rule in stock.move.line

AaronHForgeFlow reviewed (approved):
:+1:

AaronHForgeFlow commented:
/ocabot merge nobump
